<a href="https://colab.research.google.com/github/meltem12344/trendyol-e-ticaret-hackathonu-2025-kaggle/blob/main/datathon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

In [ ]:
train_raw = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/train_sessions.parquet")

In [ ]:
test_raw = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/test_sessions.parquet")

In [ ]:
test_raw.head()

,ts_hour,search_term_normalized,user_id_hashed,content_id_hashed,session_id
0,2025-07-12 17:00:00+00:00,mayo_tesettur,6ea25f8fc92c57e5,90c57f6c5e6e7359,test_0001ff614df60933
1,2025-07-12 17:00:00+00:00,mayo_tesettur,6ea25f8fc92c57e5,bfe5ebd0836137cf,test_0001ff614df60933
2,2025-07-12 17:00:00+00:00,mayo_tesettur,6ea25f8fc92c57e5,ceed8102ec139b2e,test_0001ff614df60933
3,2025-07-12 17:00:00+00:00,mayo_tesettur,6ea25f8fc92c57e5,27ccbab93a98ac37,test_0001ff614df60933
4,2025-07-12 17:00:00+00:00,mayo_tesettur,6ea25f8fc92c57e5,0f4c7995d0cd4b5f,test_0001ff614df60933


In [ ]:
test_raw["ts_hour"].value_counts()

,count
ts_hour,
2025-07-12 17:00:00+00:00,2988697


In [ ]:
train_raw.head()

,ts_hour,search_term_normalized,clicked,ordered,added_to_cart,added_to_fav,user_id_hashed,content_id_hashed,session_id
0,2025-07-10 17:00:00+00:00,a_elbise_kesim,1,0,0,0,1be4d2dad45a5e01,ae9b536d26e602f4,train_4fd3705b497bbe4f
1,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,91b8a11e4c8e0ded,train_4fd3705b497bbe4f
2,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,1a78e29f4bcc58b0,train_4fd3705b497bbe4f
3,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,604a4c8b13d798a4,train_4fd3705b497bbe4f
4,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,038680d15f693ae2,train_4fd3705b497bbe4f


In [ ]:
train_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2773805 entries, 0 to 2773804
Data columns (total 9 columns):
 #   Column                  Dtype              
---  ------                  -----              
 0   ts_hour                 datetime64[us, UTC]
 1   search_term_normalized  object             
 2   clicked                 int64              
 3   ordered                 int64              
 4   added_to_cart           int64              
 5   added_to_fav            int64              
 6   user_id_hashed          object             
 7   content_id_hashed       object             
 8   session_id              object             
dtypes: datetime64[us, UTC](1), int64(4), object(4)
memory usage: 190.5+ MB


In [ ]:
train_raw.describe()

,clicked,ordered,added_to_cart,added_to_fav
count,2.773805e+06,2.773805e+06,2.773805e+06,2.773805e+06
mean,3.576423e-02,3.077361e-03,4.054719e-03,6.233315e-03
std,1.857018e-01,5.538856e-02,6.354746e-02,7.870491e-02
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
50%,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
75%,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
max,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00


In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0


eksik değer yok
verideki satırların %3.5 i tıklanmış (imbalance)
sipariş yüzdesi %0.37    
iki aşamalı modelleme gerekiyor

In [ ]:
import pandas as pd

# 1) Toplam satır sayısı
n_rows = len(train_raw)

# 2) Click ve Order oranları
click_rate  = train_raw["clicked"].mean()
order_rate  = train_raw["ordered"].mean()

# 3) Click -> Order dönüşüm oranı
# Yalnızca click edilen ürünler içinde order edilenlerin oranı
click_to_order_ratio = train_raw["ordered"].sum() / train_raw["clicked"].sum()

print(f"Toplam satır sayısı: {n_rows:,}")
print(f"Clicked oranı:  {click_rate:.4%}")
print(f"Ordered oranı:  {order_rate:.4%}")
print(f"Click → Order dönüşüm oranı: {click_to_order_ratio:.2%}")


Toplam satır sayısı: 2,773,805
Clicked oranı:  3.5764%
Ordered oranı:  0.3077%
Click → Order dönüşüm oranı: 8.60%


📌 Bu çıktılar bize şunu söylüyor:

Click modeli zaten dengesiz (%3.5).

Order modeli çok daha zor (%0.038).

İki aşamalı modelleme (Click + Order|Click) bu yüzden şart.

In [ ]:
# Session bazlı özet
session_stats = (
    train_raw.groupby("session_id")
    .agg(
        n_items   = ("content_id_hashed", "count"),
        n_clicks  = ("clicked", "sum"),
        n_orders  = ("ordered", "sum"),
    )
    .reset_index()
)

print(session_stats.describe())

# Kaç oturum var?
print("Toplam session sayısı:", session_stats.shape[0])

# Click içeren oturum oranı
click_sessions = (session_stats["n_clicks"] > 0).mean()
print(f"Click içeren oturum oranı: {click_sessions:.2%}")

# Order içeren oturum oranı
order_sessions = (session_stats["n_orders"] > 0).mean()
print(f"Order içeren oturum oranı: {order_sessions:.2%}")

# Ortalama oturum başına click ve order sayısı
avg_clicks = session_stats["n_clicks"].mean()
avg_orders = session_stats["n_orders"].mean()
print(f"Oturum başına ortalama click sayısı: {avg_clicks:.2f}")
print(f"Oturum başına ortalama order sayısı: {avg_orders:.2f}")


            n_items      n_clicks      n_orders
count  21802.000000  21802.000000  21802.000000
mean     127.227089      4.550179      0.391524
std      185.207033      5.084884      0.679702
min       25.000000      1.000000      0.000000
25%       42.000000      2.000000      0.000000
50%       70.000000      3.000000      0.000000
75%      137.000000      6.000000      1.000000
max     7607.000000     90.000000     11.000000
Toplam session sayısı: 21802
Click içeren oturum oranı: 100.00%
Order içeren oturum oranı: 31.60%
Oturum başına ortalama click sayısı: 4.55
Oturum başına ortalama order sayısı: 0.39


1. Oturum Boyutu

Ortalama oturumda 127 ürün var (n_items.mean ≈ 127).

Ama dağılım çok geniş: minimum 25, maksimum 7607 ürün.
→ Demek ki bazı oturumlar çok kısa, bazıları aşırı uzun (muhtemelen popüler aramalarda).
→ Feature adayı: oturum uzunluğu (kaç ürün var).

In [ ]:
# User bazlı özet
user_stats = (
    train_raw.groupby("user_id_hashed")
    .agg(
        n_sessions = ("session_id", "nunique"),
        n_clicks   = ("clicked", "sum"),
        n_orders   = ("ordered", "sum"),
    )
    .reset_index()
)

print(user_stats.describe())

# Toplam kullanıcı sayısı
print("Toplam kullanıcı sayısı:", user_stats.shape[0])

# Order yapan kullanıcı oranı
order_users = (user_stats["n_orders"] > 0).mean()
print(f"En az 1 sipariş yapan kullanıcı oranı: {order_users:.2%}")

# Click yapan kullanıcı oranı
click_users = (user_stats["n_clicks"] > 0).mean()
print(f"En az 1 tıklama yapan kullanıcı oranı: {click_users:.2%}")

# Ortalama kullanıcı başına session, click, order
avg_sessions = user_stats["n_sessions"].mean()
avg_clicks   = user_stats["n_clicks"].mean()
avg_orders   = user_stats["n_orders"].mean()

print(f"Kullanıcı başına ortalama session sayısı: {avg_sessions:.2f}")
print(f"Kullanıcı başına ortalama click sayısı: {avg_clicks:.2f}")
print(f"Kullanıcı başına ortalama order sayısı: {avg_orders:.2f}")


         n_sessions      n_clicks      n_orders
count  20996.000000  20996.000000  20996.000000
mean       1.038388      4.724852      0.406554
std        0.208312      5.350187      0.743749
min        1.000000      1.000000      0.000000
25%        1.000000      2.000000      0.000000
50%        1.000000      3.000000      0.000000
75%        1.000000      6.000000      1.000000
max        4.000000     90.000000     13.000000
Toplam kullanıcı sayısı: 20996
En az 1 sipariş yapan kullanıcı oranı: 30.94%
En az 1 tıklama yapan kullanıcı oranı: 100.00%
Kullanıcı başına ortalama session sayısı: 1.04
Kullanıcı başına ortalama click sayısı: 4.72
Kullanıcı başına ortalama order sayısı: 0.41


# **Ürün verisi : metadata**

In [ ]:
metadata = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/content/metadata.parquet")

In [ ]:
metadata.head()

,level1_category_name,level2_category_name,leaf_category_name,attribute_type_count,total_attribute_option_count,merchant_count,filterable_label_count,content_creation_date,cv_tags,content_id_hashed
0,Ayakkabı,Günlük Ayakkabı,Casual Ayakkabı,1.0,6.0,1.0,4.0,2024-03-01 00:00:00+00:00,None,8dd1dce889d0e207
1,Giyim,Üst Giyim,T-Shirt,1.0,8.0,2.0,1.0,2022-03-25 00:00:00+00:00,"sokak şıklığı, rahat kesim, sokak modası",11e60da7c60ab5ea
2,Aksesuar,Takı & Mücevher,Bijuteri Kolye,1.0,1.0,1.0,0.0,2024-08-24 00:00:00+00:00,None,7562f0d2c8889460
3,Aksesuar,Gözlük,Güneş Gözlüğü,1.0,1.0,1.0,0.0,2025-07-03 00:00:00+00:00,"güneş gözlüğü, şık, ramones, göz alıcı",4c0a5002398cc709
4,Giyim,Üst Giyim,Softshell & Polar,1.0,6.0,2.0,12.0,2020-10-06 00:00:00+00:00,"hafif, daha büyük bireyler, kışın kat kat giyi...",41c07524d0266f79


In [ ]:
metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5414049 entries, 0 to 5414048
Data columns (total 10 columns):
 #   Column                        Dtype              
---  ------                        -----              
 0   level1_category_name          object             
 1   level2_category_name          object             
 2   leaf_category_name            object             
 3   attribute_type_count          float64            
 4   total_attribute_option_count  float64            
 5   merchant_count                float64            
 6   filterable_label_count        float64            
 7   content_creation_date         datetime64[us, UTC]
 8   cv_tags                       object             
 9   content_id_hashed             object             
dtypes: datetime64[us, UTC](1), float64(4), object(5)
memory usage: 413.1+ MB


In [ ]:
metadata["content_id_hashed"].nunique()

5414049



```
                yapılacaklar
1.   kategori kolonlarından hangı kategorinin kaç tıklanma sipariş aldığını çıkart
2.   bir oturumda kaç kategoriye bakılmış bunun sayısını çıkart
3.   ürün yaşını çıkart
4.   bir ürün için kaç tane tag çıkartılmış


```



In [ ]:
train_raw  = train_raw.merge(metadata, on="content_id_hashed", how="left")

In [ ]:
test_raw = test_raw.merge(metadata, on="content_id_hashed", how="left")

In [ ]:
train_raw.shape

(2773805, 18)

In [ ]:
test_raw.shape

(2988697, 14)

In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,65


boş değerleri Unknown olarak dolduralım

In [ ]:
cat_cols = ["level1_category_name", "level2_category_name", "leaf_category_name"]
for col in cat_cols:
    train_raw[col] = train_raw[col].fillna("Unknown")

In [ ]:
# Sayısal sütunlar için
num_cols = ["attribute_type_count", "total_attribute_option_count", "merchant_count", "filterable_label_count"]
for col in num_cols:
    train_raw[col] = train_raw[col].fillna(0)

# Tarih sütunu için
train_raw["content_creation_date"] = train_raw["content_creation_date"].fillna(pd.Timestamp("2000-01-01"))


In [ ]:
train_raw["cv_tags"] = train_raw["cv_tags"].fillna("Unknown")

In [ ]:
cat_cols = ["level1_category_name", "level2_category_name", "leaf_category_name"]
for col in cat_cols:
    test_raw[col] = test_raw[col].fillna("Unknown")

In [ ]:
# Sayısal sütunlar için
num_cols = ["attribute_type_count", "total_attribute_option_count", "merchant_count", "filterable_label_count"]
for col in num_cols:
    test_raw[col] = test_raw[col].fillna(0)

# Tarih sütunu için
test_raw["content_creation_date"] = test_raw["content_creation_date"].fillna(pd.Timestamp("2000-01-01"))


In [ ]:
test_raw["cv_tags"] = test_raw["cv_tags"].fillna("Unknown")

In [ ]:
test_raw.shape

(2988697, 14)

In [ ]:
train_raw.shape

(2773805, 18)

### 1.   kategori kolonlarından hangı kategorinin kaç tıklanma sipariş aldığını çıkart


In [ ]:
# leaf_category_name'e göre tıklanma sayısı
clicks_per_category = (
    train_raw.groupby("leaf_category_name")["clicked"]
      .sum()
      .reset_index()
      .rename(columns={"clicked": "leaf_category_click_count"})
)

# merge et
train_raw = train_raw.merge(clicks_per_category, on="leaf_category_name", how="left")

train_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2773805 entries, 0 to 2773804
Data columns (total 19 columns):
 #   Column                        Dtype              
---  ------                        -----              
 0   ts_hour                       datetime64[us, UTC]
 1   search_term_normalized        object             
 2   clicked                       int64              
 3   ordered                       int64              
 4   added_to_cart                 int64              
 5   added_to_fav                  int64              
 6   user_id_hashed                object             
 7   content_id_hashed             object             
 8   session_id                    object             
 9   level1_category_name          object             
 10  level2_category_name          object             
 11  leaf_category_name            object             
 12  attribute_type_count          float64            
 13  total_attribute_option_count  float64            
 14  me

In [ ]:
# merge et
test_raw = test_raw.merge(clicks_per_category, on="leaf_category_name", how="left")

In [ ]:
train_raw.shape

(2773805, 19)

In [ ]:
test_raw.shape

(2988697, 15)

In [ ]:
# --- 2) Level2 kategoriye göre toplam tıklama ---
level2_clicks = (
    train_raw.groupby("level2_category_name")["clicked"]
      .sum()
      .reset_index()
      .rename(columns={"clicked": "level2_category_click_count"})
)
train_raw = train_raw.merge(level2_clicks, on="level2_category_name", how="left")
test_raw = test_raw.merge(level2_clicks, on="level2_category_name", how="left")


# --- 3) Level1 kategoriye göre toplam tıklama ---
level1_clicks = (
    train_raw.groupby("level1_category_name")["clicked"]
      .sum()
      .reset_index()
      .rename(columns={"clicked": "level1_category_click_count"})
)
train_raw = train_raw.merge(level1_clicks, on="level1_category_name", how="left")
test_raw = test_raw.merge(level1_clicks, on="level1_category_name", how="left")


In [ ]:
train_raw.shape

(2773805, 21)

In [ ]:
test_raw.shape

(2988697, 17)

### 2.   bir oturumda kaç kategoriye bakılmış bunun sayısını çıkart


In [ ]:
# Her session'da kaç farklı kategoriye tıklanmış
session_category_counts = (
    train_raw[train_raw["clicked"] == 1]   # sadece tıklananlar
      .groupby("session_id")["leaf_category_name"]
      .nunique()
      .reset_index()
      .rename(columns={"leaf_category_name": "session_unique_category_count"})
)

# Orijinal veriye ekle
train_raw = train_raw.merge(session_category_counts, on="session_id", how="left")

In [ ]:
test_raw = test_raw.merge(session_category_counts, on="session_id", how="left")

In [ ]:
train_raw.shape

(2773805, 22)

In [ ]:
test_raw.shape

(2988697, 18)

### 3.   ürün yaşını çıkart


In [ ]:
# Emin olmak için datetime'a çeviriyoruz
train_raw["ts_hour"] = pd.to_datetime(train_raw["ts_hour"], errors="coerce", utc=True)
train_raw["content_creation_date"] = pd.to_datetime(train_raw["content_creation_date"], errors="coerce", utc=True)

# Ürün yaşını gün cinsinden hesapla
train_raw["product_age_days"] = (train_raw["ts_hour"] - train_raw["content_creation_date"]).dt.days

# Negatif değerleri sıfıra kırp (bazı ürünlerin gelecekteki tarihlerle kaydı olabilir)
train_raw["product_age_days"] = train_raw["product_age_days"].clip(lower=0)


In [ ]:
# Emin olmak için datetime'a çeviriyoruz
test_raw["ts_hour"] = pd.to_datetime(test_raw["ts_hour"], errors="coerce", utc=True)
test_raw["content_creation_date"] = pd.to_datetime(test_raw["content_creation_date"], errors="coerce", utc=True)

#   Ürün yaşını gün cinsinden hesapla
test_raw["product_age_days"] = (test_raw["ts_hour"] - test_raw["content_creation_date"]).dt.days

# Negatif değerleri sıfıra kırp (bazı ürünlerin gelecekteki tarihlerle kaydı olabilir)
test_raw["product_age_days"] = test_raw["product_age_days"].clip(lower=0)


In [ ]:
train_raw.shape

(2773805, 23)

In [ ]:
test_raw.shape

(2988697, 19)

In [ ]:
train_raw["product_age_days"].describe()

,product_age_days
count,2.773805e+06
mean,5.571566e+02
std,8.771985e+02
min,0.000000e+00
25%,1.100000e+02
50%,3.730000e+02
75%,7.530000e+02
max,9.322000e+03


### 4.   bir ürün için kaç tane tag çıkartılmış


In [ ]:
train_raw["cv_tags"]

,cv_tags
0,Unknown
1,"canlı, tatil yeri, okul, güneş"
2,"boho şık, hayvan baskısı, baskılı, güneş, sulu..."
3,"güneş, kıvrımlı, parti kıyafeti"
4,"çiçekli, canlı, çiçeksi, güneş, gül, tatil yer..."
...,...
2773800,Unknown
2773801,Unknown
2773802,"nötr, yeşil ve altın"
2773803,güneş gözlüğü


In [ ]:
def count_tags(x):
    if pd.isna(x) or x == "Unknown":
        return 0
    return len([t for t in x.split(",") if t.strip() != ""])

train_raw["cv_tag_count"] = train_raw["cv_tags"].apply(count_tags)


In [ ]:
train_raw.head()

,ts_hour,search_term_normalized,clicked,ordered,added_to_cart,added_to_fav,user_id_hashed,content_id_hashed,session_id,level1_category_name,...,merchant_count,filterable_label_count,content_creation_date,cv_tags,leaf_category_click_count,level2_category_click_count,level1_category_click_count,session_unique_category_count,product_age_days,cv_tag_count
0,2025-07-10 17:00:00+00:00,a_elbise_kesim,1,0,0,0,1be4d2dad45a5e01,ae9b536d26e602f4,train_4fd3705b497bbe4f,Giyim,...,0.0,3.0,2025-02-20 00:00:00+00:00,Unknown,11142,11142,63970,1,140,0
1,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,91b8a11e4c8e0ded,train_4fd3705b497bbe4f,Giyim,...,1.0,5.0,2025-04-03 00:00:00+00:00,"canlı, tatil yeri, okul, güneş",11142,11142,63970,1,98,4
2,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,1a78e29f4bcc58b0,train_4fd3705b497bbe4f,Giyim,...,1.0,5.0,2025-06-30 00:00:00+00:00,"boho şık, hayvan baskısı, baskılı, güneş, sulu...",11142,11142,63970,1,10,11
3,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,604a4c8b13d798a4,train_4fd3705b497bbe4f,Giyim,...,1.0,3.0,2025-02-07 00:00:00+00:00,"güneş, kıvrımlı, parti kıyafeti",11142,11142,63970,1,153,3
4,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,038680d15f693ae2,train_4fd3705b497bbe4f,Giyim,...,1.0,2.0,2025-06-10 00:00:00+00:00,"çiçekli, canlı, çiçeksi, güneş, gül, tatil yer...",11142,11142,63970,1,30,10


In [ ]:
test_raw["cv_tag_count"] = test_raw["cv_tags"].apply(count_tags)

In [ ]:
train_raw.shape

(2773805, 24)

In [ ]:
test_raw.shape

(2988697, 20)

# **Ürün verisi : price_data**

In [ ]:
price_data = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/content/price_rate_review_data.parquet")

In [ ]:
price_data.head()

,update_date,original_price,selling_price,discounted_price,content_review_count,content_review_wth_media_count,content_rate_count,content_rate_avg,content_id_hashed
0,2025-07-08,0.009995,0.009995,0.009995,9.581505e-07,4.790753e-07,9.581505e-07,5.000000,464b9f2bff5bc2a5
1,2025-07-12,0.000696,0.000696,0.000696,0.000000e+00,0.000000e+00,4.790753e-07,1.000000,9c3b5311c94101b9
2,2025-07-12,0.000511,0.000511,0.000511,3.832602e-05,1.101873e-05,5.700996e-05,3.991597,ad2805621c92d7a2
3,2025-07-12,0.000895,0.000895,0.000895,0.000000e+00,0.000000e+00,4.790753e-07,4.000000,e1606424c6933bb7
4,2025-07-08,0.001618,0.001668,0.001618,0.000000e+00,0.000000e+00,0.000000e+00,NaN,6642a900586b0107


In [ ]:
price_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1503373 entries, 0 to 1503372
Data columns (total 9 columns):
 #   Column                          Non-Null Count    Dtype         
---  ------                          --------------    -----         
 0   update_date                     1503373 non-null  datetime64[ms]
 1   original_price                  1503373 non-null  float64       
 2   selling_price                   1503373 non-null  float64       
 3   discounted_price                1503373 non-null  float64       
 4   content_review_count            1503373 non-null  float64       
 5   content_review_wth_media_count  1503373 non-null  float64       
 6   content_rate_count              1503373 non-null  float64       
 7   content_rate_avg                968578 non-null   float64       
 8   content_id_hashed               1503373 non-null  object        
dtypes: datetime64[ms](1), float64(7), object(1)
memory usage: 103.2+ MB


In [ ]:
price_data.describe()

,update_date,original_price,selling_price,discounted_price,content_review_count,content_review_wth_media_count,content_rate_count,content_rate_avg
count,1503373,1.503373e+06,1.503373e+06,1.503373e+06,1.503373e+06,1.503373e+06,1.503373e+06,968578.000000
mean,2025-07-09 14:16:34.670000,1.949897e-03,1.944824e-03,1.908041e-03,2.159103e-05,6.291594e-06,3.443708e-05,4.327073
min,2025-07-06 00:00:00,0.000000e+00,1.240006e-06,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.000000
25%,2025-07-08 00:00:00,3.712917e-04,3.662917e-04,3.562017e-04,0.000000e+00,0.000000e+00,0.000000e+00,4.043478
50%,2025-07-10 00:00:00,7.418035e-04,7.313034e-04,7.112933e-04,4.790753e-07,0.000000e+00,9.581505e-07,4.500000
75%,2025-07-12 00:00:00,1.535307e-03,1.496307e-03,1.496207e-03,4.311677e-06,9.581505e-07,6.707054e-06,4.875000
max,2025-07-12 00:00:00,1.000000e+00,1.000000e+00,1.000000e+00,3.345958e-02,1.221115e-02,5.117482e-02,5.000000
std,NaN,9.476261e-03,9.426934e-03,9.401442e-03,1.847758e-04,5.760424e-05,2.943265e-04,0.756423


In [ ]:
price_data["content_id_hashed"].nunique()

946213



```
                yapılacaklar
1.   indirim oranlarını oluştur
2.   görsel review oranını cıkart
3.   kullanılmayacak kolonları drop et
4.   content_rate_avg kolonunda 500k lık eksik var handle it
5.   tekilleştirmeyi yap


```



### 1.   indirim oranlarını oluştur


In [ ]:
import numpy as np
# (1) Satışa göre indirim oranı
price_data["discount_rate_vs_selling"] = np.where(
    price_data["selling_price"].notna(),
    (price_data["selling_price"] - price_data["discounted_price"]) / price_data["selling_price"],
    np.nan
)

# (2) Orijinale göre indirim oranı
price_data["discount_rate_vs_original"] = np.where(
    price_data["original_price"].notna(),
    (price_data["original_price"] - price_data["discounted_price"]) / price_data["original_price"],
    np.nan
)


### 2.   görsel review oranını cıkart


In [ ]:
# Görsel yorum oranı için yeni kolon
price_data["visual_review_ratio"] = price_data["content_review_wth_media_count"] / price_data["content_rate_count"]

# 0’a bölme hatalarını önlemek için sonsuz/NaN değerleri 0 yapalım
price_data["visual_review_ratio"] = price_data["visual_review_ratio"].fillna(0)


In [ ]:
price_data.head()

,update_date,original_price,selling_price,discounted_price,content_review_count,content_review_wth_media_count,content_rate_count,content_rate_avg,content_id_hashed,discount_rate_vs_selling,discount_rate_vs_original,visual_review_ratio
0,2025-07-08,0.009995,0.009995,0.009995,9.581505e-07,4.790753e-07,9.581505e-07,5.000000,464b9f2bff5bc2a5,0.000000,0.0,0.500000
1,2025-07-12,0.000696,0.000696,0.000696,0.000000e+00,0.000000e+00,4.790753e-07,1.000000,9c3b5311c94101b9,0.000000,0.0,0.000000
2,2025-07-12,0.000511,0.000511,0.000511,3.832602e-05,1.101873e-05,5.700996e-05,3.991597,ad2805621c92d7a2,0.000000,0.0,0.193277
3,2025-07-12,0.000895,0.000895,0.000895,0.000000e+00,0.000000e+00,4.790753e-07,4.000000,e1606424c6933bb7,0.000000,0.0,0.000000
4,2025-07-08,0.001618,0.001668,0.001618,0.000000e+00,0.000000e+00,0.000000e+00,NaN,6642a900586b0107,0.029971,0.0,0.000000


### 3.   kullanılmayacak kolonları drop et


In [ ]:
price_data = price_data.drop(columns=[
    "update_date",
    "selling_price",
    "discounted_price",
    "content_review_wth_media_count"
])

 ### 4.   content_rate_avg kolonunda 500k lık eksik var handle it


In [ ]:
# Ürün bazında ortalama değerleri hesapla
price_data["content_rate_avg"] = price_data.groupby("content_id_hashed")["content_rate_avg"]\
                           .transform(lambda x: x.fillna(x.mean()))


In [ ]:
price_data.isnull().sum() # bırcogunda hıc rate_avg olamdıgı ıcın hala cok fazla eksik var bunların hepsi bır kez var verıde yani , train ıle merge ettıkten sonra bakacagım

,0
original_price,0
content_review_count,0
content_rate_count,0
content_rate_avg,527154
content_id_hashed,0
discount_rate_vs_selling,0
discount_rate_vs_original,1
visual_review_ratio,0


### 5.   tekilleştirmeyi yap


In [ ]:
price_data.shape

(1503373, 8)

In [ ]:
price_data["content_id_hashed"].nunique()

946213

In [ ]:
price_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1503373 entries, 0 to 1503372
Data columns (total 8 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   original_price             1503373 non-null  float64
 1   content_review_count       1503373 non-null  float64
 2   content_rate_count         1503373 non-null  float64
 3   content_rate_avg           976219 non-null   float64
 4   content_id_hashed          1503373 non-null  object 
 5   discount_rate_vs_selling   1503373 non-null  float64
 6   discount_rate_vs_original  1503372 non-null  float64
 7   visual_review_ratio        1503373 non-null  float64
dtypes: float64(7), object(1)
memory usage: 91.8+ MB


In [ ]:
# 2) Ürün bazında tekilleştir
price_agg = (
    price_data.groupby("content_id_hashed", as_index=False)
              .mean(numeric_only=True)
)

# 3) Train ile birleştir (satır sayısı büyümez)
train_raw = train_raw.merge(price_agg, on="content_id_hashed", how="left", suffixes=("", "_price"))


In [ ]:
test_raw = test_raw.merge(price_agg, on="content_id_hashed", how="left", suffixes=("", "_price"))


In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


In [ ]:
train_raw = train_raw.drop(columns=["session_unique_category_count"])
test_raw = test_raw.drop(columns=["session_unique_category_count"])

boşları dolduralım

In [ ]:
# Boş olan kolonlar
cols_to_fill = [
    "original_price",
    "content_review_count",
    "content_rate_count",
    "content_rate_avg",
    "discount_rate_vs_selling",
    "discount_rate_vs_original",
    "visual_review_ratio"
]


for c in cols_to_fill:
    # Eksik bayrağı ekle
    train_raw[f"{c}_was_null"] = train_raw[c].isna().astype(np.int8)

    # Leaf kategori bazında medyanı hesapla
    med_leaf = train_raw.groupby("leaf_category_name")[c].transform("median")

    # Önce leaf medyan ile doldur
    tmp = train_raw[c].fillna(med_leaf)

    # Hâlâ boş kalanları genel medyanla doldur
    train_raw[c] = tmp.fillna(train_raw[c].median())

In [ ]:
# Boş olan kolonlar
cols_to_fill = [
    "original_price",
    "content_review_count",
    "content_rate_count",
    "content_rate_avg",
    "discount_rate_vs_selling",
    "discount_rate_vs_original",
    "visual_review_ratio"
]


for c in cols_to_fill:
    # Eksik bayrağı ekle
    test_raw[f"{c}_was_null"] = test_raw[c].isna().astype(np.int8)

    # Leaf kategori bazında medyanı hesapla
    med_leaf = test_raw.groupby("leaf_category_name")[c].transform("median")

    # Önce leaf medyan ile doldur
    tmp = test_raw[c].fillna(med_leaf)

    # Hâlâ boş kalanları genel medyanla doldur
    test_raw[c] = tmp.fillna(test_raw[c].median())

In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


In [ ]:
train_raw.shape

(2773805, 37)

In [ ]:
test_raw.shape

(2988697, 33)

### meraktan

In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


In [ ]:
train_raw["discount_rate_vs_selling"].value_counts()

,count
discount_rate_vs_selling,
0.000000,2534843
0.075393,1342
0.075710,1244
0.100324,1129
0.075442,834
...,...
0.045124,1
0.050137,1
0.050116,1


In [ ]:
# İndirim oranı sıfırdan büyük olanları filtrele
indirimli = train_raw[train_raw["discount_rate_vs_selling"] > 0]

# Kaç tanesi tıklanmış?
tiklanan = indirimli["clicked"].sum()

# Kaç tanesi sipariş edilmiş?
siparis_edilen = indirimli["ordered"].sum()

print("Tıklanan ürün sayısı:", tiklanan)
print("Sipariş edilen ürün sayısı:", siparis_edilen)


Tıklanan ürün sayısı: 8622
Sipariş edilen ürün sayısı: 858


In [ ]:
train_raw["discount_rate_vs_original"].value_counts()

,count
discount_rate_vs_original,
0.000000,2067559
0.203087,1904
0.214391,1865
0.287888,1665
0.185529,1503
...,...
0.315769,1
0.082342,1
0.501859,1


In [ ]:
train_raw["session_id"].nunique()

21802

# **Ürün verisi : search_log**

In [ ]:
search_log = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/content/search_log.parquet")

In [ ]:
search_log.head()

,date,total_search_impression,total_search_click,content_id_hashed
0,2025-06-15,0.000059,4.790753e-07,8e61a30fa370af02
1,2025-06-30,0.000001,0.000000e+00,22b7d2a132d9332b
2,2025-06-19,0.000147,3.832602e-06,c0236381c4dfc7f5
3,2025-07-07,0.000024,0.000000e+00,b36bcbd981cabfcd
4,2025-06-30,0.000006,0.000000e+00,c76ff84c6c16d786


In [ ]:
search_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32128237 entries, 0 to 32128236
Data columns (total 4 columns):
 #   Column                   Dtype         
---  ------                   -----         
 0   date                     datetime64[ms]
 1   total_search_impression  float64       
 2   total_search_click       float64       
 3   content_id_hashed        object        
dtypes: datetime64[ms](1), float64(2), object(1)
memory usage: 980.5+ MB




```
                yapılacaklar
1.   ctr hesapla
2.   popülerlik skoru hesapla
3.   kullanılmayacak kolonları drop et
4.   tekilleştir


```



### 1.   ctr hesapla


In [ ]:
# CTR hesapla
search_log["ctr"] = search_log["total_search_click"] / search_log["total_search_impression"]

In [ ]:
import numpy as np

search_log["ctr"] = np.where(
    search_log["total_search_impression"] > 0,
    search_log["total_search_click"] / search_log["total_search_impression"],
    0
)

### 2.   popülerlik skoru hesapla


In [ ]:
# Normalize edilmiş popülerlik skoru
total_impressions = search_log["total_search_impression"].sum()
search_log["popularity_score"] = search_log["total_search_impression"] / total_impressions


### 3.   kullanılmayacak kolonları drop et


In [ ]:
search_log.head()

,date,total_search_impression,total_search_click,content_id_hashed,ctr,popularity_score
0,2025-06-15,0.000059,4.790753e-07,8e61a30fa370af02,0.008130,7.564159e-09
1,2025-06-30,0.000001,0.000000e+00,22b7d2a132d9332b,0.000000,1.844917e-10
2,2025-06-19,0.000147,3.832602e-06,c0236381c4dfc7f5,0.026144,1.881815e-08
3,2025-07-07,0.000024,0.000000e+00,b36bcbd981cabfcd,0.000000,3.136359e-09
4,2025-06-30,0.000006,0.000000e+00,c76ff84c6c16d786,0.000000,7.994640e-10


In [ ]:
search_log.drop(columns=["total_search_impression", "total_search_click"], inplace=True)

### 4.   tekilleştir

In [ ]:
search_log["content_id_hashed"].nunique()

1161712

In [ ]:
# content_id_hashed bazında tekilleştir
search_log = (
    search_log
    .groupby("content_id_hashed", as_index=False)
    .agg({
        "ctr": "mean",              # ortalama CTR
        "popularity_score": "sum",  # popülerlik skorlarını toplayabilirsin
        "date": "min"               # en eski tarihi al (ya da "max" yapabilirsin)
    })
)

In [ ]:
train_raw = train_raw.merge(search_log, on="content_id_hashed", how="left")

In [ ]:
train_raw.shape

(2773805, 40)

In [ ]:
test_raw = test_raw.merge(search_log, on="content_id_hashed", how="left")

In [ ]:
test_raw.shape

(2988697, 36)

In [ ]:
train_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2773805 entries, 0 to 2773804
Data columns (total 40 columns):
 #   Column                              Dtype              
---  ------                              -----              
 0   ts_hour                             datetime64[us, UTC]
 1   search_term_normalized              object             
 2   clicked                             int64              
 3   ordered                             int64              
 4   added_to_cart                       int64              
 5   added_to_fav                        int64              
 6   user_id_hashed                      object             
 7   content_id_hashed                   object             
 8   session_id                          object             
 9   level1_category_name                object             
 10  level2_category_name                object             
 11  leaf_category_name                  object             
 12  attribute_type_count        

In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


boşları dolduralım

In [ ]:
# 1) Eksik bayrakları
for c in ["ctr", "popularity_score", "date"]:
    test_raw[f"{c}_was_null"] = test_raw[c].isna().astype(np.int8)

# 2) Sayısallar: ctr & popularity_score -> ürün (content_id_hashed) medyanı
for c in ["ctr", "popularity_score"]:
    grp_med = test_raw.groupby("content_id_hashed")[c].transform("median")
    test_raw[c] = test_raw[c].fillna(grp_med) \
                 .fillna(test_raw[c].median())  # hâlâ NaN ise global medyan

# 3) Tarih: medyan tarih ile doldur (datetime’a çevirip int üzerinden medyan)
test_raw["date"] = pd.to_datetime(test_raw["date"], errors="coerce")

# content_id_hashed bazında medyan tarih (ns int -> geri datetime)
date_int = test_raw["date"].astype("int64")  # NaT -> NaN kalır
grp_med_date_int = date_int.groupby(test_raw["content_id_hashed"]).transform("median")
grp_med_date = pd.to_datetime(grp_med_date_int)

# doldur: önce ürün medyanı, sonra global medyan
global_med_date = pd.to_datetime(date_int.median())  # tüm set medyan tarihi
test_raw["date"] = test_raw["date"].fillna(grp_med_date).fillna(global_med_date)

In [ ]:
# 1) Eksik bayrakları
for c in ["ctr", "popularity_score", "date"]:
    train_raw[f"{c}_was_null"] = train_raw[c].isna().astype(np.int8)

# 2) Sayısallar: ctr & popularity_score -> ürün (content_id_hashed) medyanı
for c in ["ctr", "popularity_score"]:
    grp_med = train_raw.groupby("content_id_hashed")[c].transform("median")
    train_raw[c] = train_raw[c].fillna(grp_med) \
                 .fillna(train_raw[c].median())  # hâlâ NaN ise global medyan

# 3) Tarih: medyan tarih ile doldur (datetime’a çevirip int üzerinden medyan)
train_raw["date"] = pd.to_datetime(train_raw["date"], errors="coerce")

# content_id_hashed bazında medyan tarih (ns int -> geri datetime)
date_int = train_raw["date"].astype("int64")  # NaT -> NaN kalır
grp_med_date_int = date_int.groupby(train_raw["content_id_hashed"]).transform("median")
grp_med_date = pd.to_datetime(grp_med_date_int)

# doldur: önce ürün medyanı, sonra global medyan
global_med_date = pd.to_datetime(date_int.median())  # tüm set medyan tarihi
train_raw["date"] = train_raw["date"].fillna(grp_med_date).fillna(global_med_date)

In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


# **Ürün verisi : sitewide_log**

In [ ]:
sitewide_log = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/content/sitewide_log.parquet")

In [ ]:
sitewide_log.head()

,date,total_click,total_cart,total_fav,total_order,content_id_hashed
0,2025-06-30,0.000114,3.353527e-06,1.437226e-05,9.581505e-07,caf9228a3c5c8d13
1,2025-06-30,0.000002,0.000000e+00,0.000000e+00,0.000000e+00,351d178f1567ebb0
2,2025-06-28,0.000008,0.000000e+00,9.581505e-07,0.000000e+00,de124335a4ec02e2
3,2025-07-03,0.000002,0.000000e+00,4.790753e-07,0.000000e+00,c38da9ee0d8c5a93
4,2025-07-04,0.000041,9.581505e-07,6.707054e-06,9.581505e-07,623730ad5fe077c0


In [ ]:
sitewide_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23977263 entries, 0 to 23977262
Data columns (total 6 columns):
 #   Column             Dtype         
---  ------             -----         
 0   date               datetime64[ms]
 1   total_click        float64       
 2   total_cart         float64       
 3   total_fav          float64       
 4   total_order        float64       
 5   content_id_hashed  object        
dtypes: datetime64[ms](1), float64(4), object(1)
memory usage: 1.1+ GB


In [ ]:
sitewide_log.drop(columns=["date"], inplace=True)

In [ ]:
sitewide_log["content_id_hashed"].nunique()

1118805

In [ ]:
sitewide_log.isnull().sum()

,0
total_click,0
total_cart,0
total_fav,0
total_order,0
content_id_hashed,0


In [ ]:
sitewide_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23977263 entries, 0 to 23977262
Data columns (total 5 columns):
 #   Column             Dtype  
---  ------             -----  
 0   total_click        float64
 1   total_cart         float64
 2   total_fav          float64
 3   total_order        float64
 4   content_id_hashed  object 
dtypes: float64(4), object(1)
memory usage: 914.7+ MB


### 1.  tekilleştir

In [ ]:
# Sayısal kolonları seç
val_cols = sitewide_log.columns.drop('content_id_hashed')

# content_id_hashed bazında tekilleştir (toplam)
sitewide_agg = (
    sitewide_log
    .groupby('content_id_hashed', as_index=False, sort=False)[val_cols]
    .sum()
)


In [ ]:
train_raw = train_raw.merge(sitewide_agg, on="content_id_hashed", how="left")
test_raw = test_raw.merge(sitewide_agg, on="content_id_hashed", how="left")

In [ ]:
print(train_raw.shape)
print(test_raw.shape)

(2773805, 47)
(2988697, 43)


In [ ]:
test_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2988697 entries, 0 to 2988696
Data columns (total 43 columns):
 #   Column                              Dtype              
---  ------                              -----              
 0   ts_hour                             datetime64[us, UTC]
 1   search_term_normalized              object             
 2   user_id_hashed                      object             
 3   content_id_hashed                   object             
 4   session_id                          object             
 5   level1_category_name                object             
 6   level2_category_name                object             
 7   leaf_category_name                  object             
 8   attribute_type_count                float64            
 9   total_attribute_option_count        float64            
 10  merchant_count                      float64            
 11  filterable_label_count              float64            
 12  content_creation_date       

In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


boşları dolduralım

In [ ]:
# Doldurulacak kolonlar
count_cols = ["total_click", "total_cart", "total_fav", "total_order"]

for c in count_cols:
    # NaN'leri medyan ile doldur
    train_raw[c] = train_raw[c].fillna(train_raw[c].median())

In [ ]:
# Doldurulacak kolonlar
count_cols = ["total_click", "total_cart", "total_fav", "total_order"]

for c in count_cols:
    # NaN'leri medyan ile doldur
    test_raw[c] = test_raw[c].fillna(test_raw[c].median())

In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


# **Ürün verisi : top_terms_log**

In [ ]:
top_terms_log = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/content/top_terms_log.parquet")

In [ ]:
top_terms_log.head()

,date,search_term_normalized,total_search_impression,total_search_click,content_id_hashed
0,2025-06-27,cocuk_kiz_mayo,2.395376e-06,0.000000e+00,90feecae99ac21ff
1,2025-07-07,kot_pantolon,1.437226e-06,4.790753e-07,ff85d4e14280f57f
2,2025-06-22,gecelik,1.159362e-04,4.311677e-06,8bea12a22a38d2e8
3,2025-06-20,canta,4.790753e-06,0.000000e+00,333fdfc7b60e5b0e
4,2025-06-14,boxer,4.790753e-07,0.000000e+00,df74bcf645a6eea3


In [ ]:
top_terms_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45743432 entries, 0 to 45743431
Data columns (total 5 columns):
 #   Column                   Dtype         
---  ------                   -----         
 0   date                     datetime64[ms]
 1   search_term_normalized   object        
 2   total_search_impression  float64       
 3   total_search_click       float64       
 4   content_id_hashed        object        
dtypes: datetime64[ms](1), float64(2), object(2)
memory usage: 1.7+ GB


In [ ]:
# --- 2) Terime bağlı alaka: top_terms_log ---
t = (top_terms_log
     .groupby(['content_id_hashed','search_term_normalized'], as_index=False)
     .agg(t_imp=('total_search_impression','sum'),
          t_clk=('total_search_click','sum')))
t['term_ctr'] = np.where(t['t_imp'] > 0, t['t_clk'] / t['t_imp'], 0.0)
t = t[['content_id_hashed','search_term_normalized','term_ctr']]


In [ ]:
train_raw = train_raw.merge(t, on=['content_id_hashed','search_term_normalized'], how='left')


In [ ]:
train_raw.shape

(2773805, 48)

In [ ]:
test_raw = test_raw.merge(t, on=['content_id_hashed','search_term_normalized'], how='left')

In [ ]:
train_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2773805 entries, 0 to 2773804
Data columns (total 48 columns):
 #   Column                              Dtype              
---  ------                              -----              
 0   ts_hour                             datetime64[us, UTC]
 1   search_term_normalized              object             
 2   clicked                             int64              
 3   ordered                             int64              
 4   added_to_cart                       int64              
 5   added_to_fav                        int64              
 6   user_id_hashed                      object             
 7   content_id_hashed                   object             
 8   session_id                          object             
 9   level1_category_name                object             
 10  level2_category_name                object             
 11  leaf_category_name                  object             
 12  attribute_type_count        

In [ ]:
test_raw.shape

(2988697, 44)

In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


boşları dolduralım

In [ ]:
# Terim CTR doldurma
if "search_term_normalized" in train_raw.columns:
    grp_med = train_raw.groupby("search_term_normalized")["term_ctr"].transform("median")
    train_raw["term_ctr"] = train_raw["term_ctr"].fillna(grp_med) \
                                               .fillna(train_raw["term_ctr"].median())
else:
    # sadece global medyan
    train_raw["term_ctr"] = train_raw["term_ctr"].fillna(train_raw["term_ctr"].median())

In [ ]:
# Terim CTR doldurma
if "search_term_normalized" in train_raw.columns:
    grp_med = test_raw.groupby("search_term_normalized")["term_ctr"].transform("median")
    test_raw["term_ctr"] = test_raw["term_ctr"].fillna(grp_med) \
                                               .fillna(test_raw["term_ctr"].median())
else:
    # sadece global medyan
    test_raw["term_ctr"] = test_raw["term_ctr"].fillna(test_raw["term_ctr"].median())

# **Kullanıcı verisi : metadata**

In [ ]:
user_metadata = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/user/metadata.parquet")

In [ ]:
user_metadata.head()

,user_gender,user_birth_year,user_tenure_in_days,user_id_hashed
0,Bayan,1992.0,2619,94238d723579f0bd
1,UNKNOWN,1998.0,1054,b041bad2ad93c114
2,Bayan,1996.0,2179,90361194b60e5f3b
3,Bayan,1979.0,1930,e570ed7193aa197a
4,UNKNOWN,1987.0,1495,f71c3c802b3934a1


In [ ]:
user_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38392 entries, 0 to 38391
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_gender          38392 non-null  object 
 1   user_birth_year      15572 non-null  float64
 2   user_tenure_in_days  38392 non-null  int64  
 3   user_id_hashed       38392 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 1.2+ MB


In [ ]:
user_metadata["user_id_hashed"].nunique()

38392

In [ ]:
user_metadata.isnull().sum()

,0
user_gender,0
user_birth_year,22820
user_tenure_in_days,0
user_id_hashed,0


In [ ]:
user_metadata["user_tenure_in_days"].describe()

,user_tenure_in_days
count,38392.000000
mean,1798.466738
std,1291.089121
min,25.000000
25%,832.000000
50%,1677.000000
75%,2337.000000
max,5638.000000


In [ ]:
user_metadata["user_birth_year"].describe()

,user_birth_year
count,15572.000000
mean,1981.445286
std,90.037215
min,1.000000
25%,1985.000000
50%,1994.000000
75%,1999.000000
max,2021.000000


In [ ]:
# Sayısal kolonlar için ortalama
num_cols = ["user_birth_year", "user_tenure_in_days"]

user_meta = (
    user_metadata
    .groupby("user_id_hashed", as_index=False)
    .agg({
        "user_gender": "first",   # her kullanıcıda sabit olduğu için ilkini al
        "user_birth_year": "mean",
        "user_tenure_in_days": "mean"
    })
)

# Eğer integer olsun istersen:
user_meta["user_birth_year"] = user_meta["user_birth_year"].round().astype("Int64")
user_meta["user_tenure_in_days"] = user_meta["user_tenure_in_days"].round().astype("Int64")


# 3) LEFT MERGE (train tarafı korunur)
train_raw = train_raw.merge(
    user_meta,
    on='user_id_hashed',
    how='left',
    validate='m:1'   # user_meta’da her user tek satır olmalı; aksi halde hata verir
)

train_raw.head()

,ts_hour,search_term_normalized,clicked,ordered,added_to_cart,added_to_fav,user_id_hashed,content_id_hashed,session_id,level1_category_name,...,popularity_score_was_null,date_was_null,total_click,total_cart,total_fav,total_order,term_ctr,user_gender,user_birth_year,user_tenure_in_days
0,2025-07-10 17:00:00+00:00,a_elbise_kesim,1,0,0,0,1be4d2dad45a5e01,ae9b536d26e602f4,train_4fd3705b497bbe4f,Giyim,...,0,0,0.008343,0.000182,0.000720,0.000039,0.022222,UNKNOWN,2003,371
1,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,91b8a11e4c8e0ded,train_4fd3705b497bbe4f,Giyim,...,0,0,0.043029,0.000832,0.004832,0.000123,0.037954,UNKNOWN,2003,371
2,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,1a78e29f4bcc58b0,train_4fd3705b497bbe4f,Giyim,...,0,0,0.006076,0.000256,0.001463,0.000042,0.029703,UNKNOWN,2003,371
3,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,604a4c8b13d798a4,train_4fd3705b497bbe4f,Giyim,...,0,0,0.019713,0.000744,0.002537,0.000137,0.015005,UNKNOWN,2003,371
4,2025-07-10 17:00:00+00:00,a_elbise_kesim,0,0,0,0,1be4d2dad45a5e01,038680d15f693ae2,train_4fd3705b497bbe4f,Giyim,...,0,0,0.037996,0.000675,0.006748,0.000133,0.018950,UNKNOWN,2003,371


In [ ]:
# 3) LEFT MERGE (test tarafı korunur)
test_raw = test_raw.merge(
    user_meta,
    on='user_id_hashed',
    how='left',
    validate='m:1'   # user_meta’da her user tek satır olmalı; aksi halde hata verir
)


In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


In [ ]:
test_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0
level2_category_name,0
leaf_category_name,0
attribute_type_count,0
total_attribute_option_count,0


In [ ]:
train_raw.drop(columns=["user_birth_year"], inplace=True)
test_raw.drop(columns=["user_birth_year"], inplace=True)

In [ ]:
train_raw.shape

(2773805, 50)

In [ ]:
test_raw.shape

(2988697, 46)

# **Kullanıcı verisi : search_log**

In [ ]:
user_search_log = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/content/search_log.parquet")

In [ ]:
user_search_log.head()

,date,total_search_impression,total_search_click,content_id_hashed
0,2025-06-15,0.000059,4.790753e-07,8e61a30fa370af02
1,2025-06-30,0.000001,0.000000e+00,22b7d2a132d9332b
2,2025-06-19,0.000147,3.832602e-06,c0236381c4dfc7f5
3,2025-07-07,0.000024,0.000000e+00,b36bcbd981cabfcd
4,2025-06-30,0.000006,0.000000e+00,c76ff84c6c16d786


In [ ]:
user_search_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32128237 entries, 0 to 32128236
Data columns (total 4 columns):
 #   Column                   Dtype         
---  ------                   -----         
 0   date                     datetime64[ms]
 1   total_search_impression  float64       
 2   total_search_click       float64       
 3   content_id_hashed        object        
dtypes: datetime64[ms](1), float64(2), object(1)
memory usage: 980.5+ MB


In [ ]:
import numpy as np
import pandas as pd

# user_search_log: [ts_hour, total_search_impression, total_search_click, user_id_hashed]
user_search_log = (user_search_log
        .groupby('user_id_hashed', as_index=False)
        .agg(user_search_imp_sum=('total_search_impression','sum'),
             user_search_clk_sum=('total_search_click','sum')))

user_search_log['user_search_ctr'] = np.where(
    user_search_log['user_search_imp_sum'] > 0,
    user_search_log['user_search_clk_sum'] / user_search_log['user_search_imp_sum'],
    0.0
).astype('float32')


KeyError: 'user_id_hashed'

In [ ]:
user_search_log.head()

In [ ]:
train_raw = train_raw.merge(
    user_search_log, on='user_id_hashed', how='left', validate='m:1'
)

In [ ]:
test_raw = test_raw.merge(
    user_search_log, on='user_id_hashed', how='left', validate='m:1'
)

In [ ]:
train_raw.shape

In [ ]:
test_raw.shape

In [ ]:
train_raw.isnull().sum()

In [ ]:
test_raw.isnull().sum()

In [ ]:
train_raw.info()

boşları dolduralım

In [ ]:
# Eksikleri medyanla doldur
search_cols = ["user_search_imp_sum", "user_search_clk_sum", "user_search_ctr"]

for c in search_cols:
    train_raw[c] = train_raw[c].fillna(train_raw[c].median())

In [ ]:
# Eksikleri medyanla doldur
search_cols = ["user_search_imp_sum", "user_search_clk_sum", "user_search_ctr"]

for c in search_cols:
    test_raw[c] = test_raw[c].fillna(test_raw[c].median())

# **Kullanıcı : sitewide_log**

In [ ]:
user_sitewide_log = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/user/sitewide_log.parquet")

In [ ]:
user_sitewide_log.head()

,ts_hour,total_click,total_cart,total_fav,total_order,user_id_hashed
0,2025-06-22 12:00:00+00:00,1.437226e-06,0.000000e+00,0.000000e+00,0.0,957161b655ba4427
1,2025-06-18 22:00:00+00:00,2.395376e-06,4.790753e-07,9.581505e-07,0.0,fe9b0f2b3c5f9404
2,2025-06-09 13:00:00+00:00,3.353527e-06,4.790753e-07,9.581505e-07,0.0,ce02e76489bc6b11
3,2025-06-30 17:00:00+00:00,4.790753e-07,0.000000e+00,0.000000e+00,0.0,d76bce6c7c731647
4,2025-06-27 21:00:00+00:00,2.874452e-06,0.000000e+00,0.000000e+00,0.0,480aa52af626fd1c


In [ ]:
user_sitewide_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1891488 entries, 0 to 1891487
Data columns (total 6 columns):
 #   Column          Dtype              
---  ------          -----              
 0   ts_hour         datetime64[us, UTC]
 1   total_click     float64            
 2   total_cart      float64            
 3   total_fav       float64            
 4   total_order     float64            
 5   user_id_hashed  object             
dtypes: datetime64[us, UTC](1), float64(4), object(1)
memory usage: 86.6+ MB


In [ ]:
# Kullanıcı bazında ortalama değerler
user_sitewide_log = (
    user_sitewide_log
    .groupby("user_id_hashed", as_index=False)
    [["total_click", "total_cart", "total_fav", "total_order"]]
    .mean()
    .rename(columns=lambda x: x + "_mean" if x != "user_id_hashed" else x)
)

In [ ]:
train_raw = train_raw.merge(
    user_sitewide_log,
    on="user_id_hashed",
    how="left",   # train’deki tüm kayıtlar kalsın, eşleşen user verisi gelsin
    suffixes=("", "_user")  # çakışan kolon olursa user tarafına _user eklensin
)

In [ ]:
train_raw.isnull().sum()

,0
ts_hour,0
search_term_normalized,0
clicked,0
ordered,0
added_to_cart,0
added_to_fav,0
user_id_hashed,0
content_id_hashed,0
session_id,0
level1_category_name,0


In [ ]:
test_raw = test_raw.merge(
    user_sitewide_log,
    on="user_id_hashed",
    how="left",   # train’deki tüm kayıtlar kalsın, eşleşen user verisi gelsin
    suffixes=("", "_user")  # çakışan kolon olursa user tarafına _user eklensin
)

In [ ]:
test_raw.shape

(2988697, 50)

In [ ]:
train_raw.shape

(2773805, 54)

In [ ]:
# 0) Güvenlik: bölmede sıfır hatasını önlemek için küçük sabit
EPS = 1e-5


# 2) Basit, işlevsel feature'lar
test_raw["u_engagement"]   = test_raw["total_click"] + test_raw["total_cart"] + test_raw["total_fav"]
test_raw["u_click_ratio"]  = test_raw["total_click"] / (test_raw["total_cart"] + test_raw["total_fav"] + test_raw["total_order"] + EPS)
test_raw["u_conv_rate"]    = test_raw["total_order"] / (test_raw["total_click"] + EPS)
test_raw["u_cart2order"]   = test_raw["total_order"] / (test_raw["total_cart"] + EPS)


In [ ]:
# 0) Güvenlik: bölmede sıfır hatasını önlemek için küçük sabit
EPS = 1e-5


# 2) Basit, işlevsel feature'lar
train_raw["u_engagement"]   = train_raw["total_click"] + train_raw["total_cart"] + train_raw["total_fav"]
train_raw["u_click_ratio"]  = train_raw["total_click"] / (train_raw["total_cart"] + train_raw["total_fav"] + train_raw["total_order"] + EPS)
train_raw["u_conv_rate"]    = train_raw["total_order"] / (train_raw["total_click"] + EPS)
train_raw["u_cart2order"]   = train_raw["total_order"] / (train_raw["total_cart"] + EPS)


In [ ]:
test_raw.shape

(2988697, 54)

In [ ]:
train_raw.shape

(2773805, 58)

In [ ]:
user_fashion_search_log = pd.read_parquet("/content/drive/MyDrive/Drive /Dataminds/trendyol-e-ticaret-hackathonu-2025-kaggle/data/user/fashion_search_log.parquet")

In [ ]:
user_fashion_search_log.head()

,ts_hour,total_search_impression,total_search_click,user_id_hashed,content_id_hashed
0,2025-07-05 12:00:00+00:00,4.790753e-07,0.000000e+00,61e5a21740349223,167c69c06a4adee8
1,2025-07-04 17:00:00+00:00,0.000000e+00,4.790753e-07,127631d27551594c,d381b1bb256d3a20
2,2025-06-15 23:00:00+00:00,4.790753e-07,0.000000e+00,9a19f5739f727764,371d8ae47e99b8ea
3,2025-07-01 14:00:00+00:00,4.790753e-07,0.000000e+00,fda8cd220a879eba,a2c6fb61103cd240
4,2025-07-07 10:00:00+00:00,4.790753e-07,0.000000e+00,086cd81a3bdbd176,72764e8e8db8fa46


In [ ]:
user_fashion_search_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103388974 entries, 0 to 103388973
Data columns (total 5 columns):
 #   Column                   Dtype              
---  ------                   -----              
 0   ts_hour                  datetime64[us, UTC]
 1   total_search_impression  float64            
 2   total_search_click       float64            
 3   user_id_hashed           object             
 4   content_id_hashed        object             
dtypes: datetime64[us, UTC](1), float64(2), object(2)
memory usage: 3.9+ GB


In [ ]:
user_fashion_search_log["user_id_hashed"].nunique()

36949

In [ ]:

# 1) Ürün bazında grupla
user_fashion_search_log = (
    user_fashion_search_log
    .groupby("content_id_hashed", as_index=False)
    .agg({
        "total_search_impression": "sum",
        "total_search_click": "sum"
    })
)

# 2) CTR hesapla
user_fashion_search_log["content_ctr"] = user_fashion_search_log["total_search_click"] / (user_fashion_search_log["total_search_impression"] + EPS)

# 3) Train ile merge et (content_id bazında)
train_raw = train_raw.merge(
    user_fashion_search_log,
    on="content_id_hashed",
    how="left",
    suffixes=("", "_content")
)

In [ ]:

# 3) test ile merge et (content_id bazında)
test_raw = test_raw.merge(
    user_fashion_search_log,
    on="content_id_hashed",
    how="left",
    suffixes=("", "_content")
)

sonra bunlar

In [ ]:
'''user_fashion_sitewide_log = pd.read_parquet("/content/drive/MyDrive/trendyol-e-ticaret-hackathonu-2025-kaggle/data/user/fashion_sitewide_log.parquet")'''

'user_fashion_sitewide_log = pd.read_parquet("/content/drive/MyDrive/trendyol-e-ticaret-hackathonu-2025-kaggle/data/user/fashion_sitewide_log.parquet")'

In [ ]:
'''user_search_log = pd.read_parquet("/content/drive/MyDrive/trendyol-e-ticaret-hackathonu-2025-kaggle/data/user/search_log.parquet")'''

'user_search_log = pd.read_parquet("/content/drive/MyDrive/trendyol-e-ticaret-hackathonu-2025-kaggle/data/user/search_log.parquet")'

In [ ]:
train_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2773805 entries, 0 to 2773804
Data columns (total 61 columns):
 #   Column                              Dtype              
---  ------                              -----              
 0   ts_hour                             datetime64[us, UTC]
 1   search_term_normalized              object             
 2   clicked                             int64              
 3   ordered                             int64              
 4   added_to_cart                       int64              
 5   added_to_fav                        int64              
 6   user_id_hashed                      object             
 7   content_id_hashed                   object             
 8   session_id                          object             
 9   level1_category_name                object             
 10  level2_category_name                object             
 11  leaf_category_name                  object             
 12  attribute_type_count        

In [ ]:
test_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2988697 entries, 0 to 2988696
Data columns (total 57 columns):
 #   Column                              Dtype              
---  ------                              -----              
 0   ts_hour                             datetime64[us, UTC]
 1   search_term_normalized              object             
 2   user_id_hashed                      object             
 3   content_id_hashed                   object             
 4   session_id                          object             
 5   level1_category_name                object             
 6   level2_category_name                object             
 7   leaf_category_name                  object             
 8   attribute_type_count                float64            
 9   total_attribute_option_count        float64            
 10  merchant_count                      float64            
 11  filterable_label_count              float64            
 12  content_creation_date       

# **Model**

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import GroupKFold
from sklearn.metrics import roc_auc_score
from lightgbm import LGBMClassifier

SEED = 42


In [ ]:
def session_auc(df, session_col, y_true_col, y_score_col):
    scores = []
    for sid, g in df.groupby(session_col):
        y = g[y_true_col].values
        if (y.sum() == 0) or (y.sum() == len(y)):
            continue
        s = roc_auc_score(y, g[y_score_col].values)
        scores.append(s)
    return float(np.mean(scores)) if scores else np.nan


In [ ]:
def prep(df, is_train=True):
    out = df.copy()

    # Tarih özellikleri
    if 'ts_hour' in out.columns:
        out['ts_hour_hour'] = out['ts_hour'].dt.hour.astype('int16')
        out['ts_hour_dow']  = out['ts_hour'].dt.dayofweek.astype('int8')
    else:
        out['ts_hour_hour'] = 0
        out['ts_hour_dow']  = 0


    # Sayısal kolonlar =
    num_cols = [
        'total_attribute_option_count',
        'merchant_count',
        'filterable_label_count',
        'original_price',
        'content_review_count',
        'content_rate_count',
        'content_rate_avg',
        'discount_rate_vs_selling',
        'discount_rate_vs_original',
        'visual_review_ratio',
        'ctr',
        'popularity_score',
        'total_click',
        'total_cart',
        'total_fav',
        'total_order',
        'term_ctr',
        'user_tenure_in_days',
        'user_search_imp_sum',
        'user_search_clk_sum',
        'user_search_ctr',
        'cv_tag_count',
        'product_age_days',
        'leaf_category_click_count', 'total_cart_mean', 'total_fav_mean',
        'total_order_mean', 'u_engagement', 'u_click_ratio', 'u_conv_rate',
        'u_cart2order','total_search_impression','total_search_click','content_ctr'
    ]


    num_cols = [c for c in num_cols if c in out.columns]

    # Kategorikler
    cat_cols = [
        'level1_category_name',
        'user_gender'
    ]
    cat_cols = [c for c in cat_cols if c in out.columns]
    for c in cat_cols:
        out[c] = out[c].astype('category')

    keep = num_cols + cat_cols + ['session_id','content_id_hashed']
    if is_train:
        targets = [c for c in ['clicked','ordered'] if c in out.columns]
        keep += targets
    return out[keep], num_cols, cat_cols

train_df, NUM_COLS, CAT_COLS = prep(train_raw, is_train=True)


In [ ]:
gkf = GroupKFold(n_splits=5)
groups = train_df['session_id'].values
FEATURES = NUM_COLS + CAT_COLS


In [ ]:
params_base = dict(
    n_estimators=1500,
    learning_rate=0.03,
    max_depth=-1,
    num_leaves=63,
    subsample=0.8,
    colsample_bytree=0.8,
    reg_alpha=1.0,
    reg_lambda=2.0,
    random_state=SEED,
    n_jobs=-1,
    device='gpu',
    gpu_platform_id=0,
    gpu_device_id=0
)

def fit_lgbm(train_df, target_col):
    oof_pred = np.zeros(len(train_df), dtype=float)
    models = []

    for fold, (tr_idx, va_idx) in enumerate(gkf.split(train_df, groups=groups)):
        tr, va = train_df.iloc[tr_idx], train_df.iloc[va_idx]

        clf = LGBMClassifier(**params_base)
        clf.fit(
            tr[FEATURES], tr[target_col],
            eval_set=[(va[FEATURES], va[target_col])],
            eval_metric='auc',
            categorical_feature=CAT_COLS
        )
        models.append(clf)
        oof_pred[va_idx] = clf.predict_proba(va[FEATURES])[:,1]

        va_tmp = va[['session_id', target_col]].copy()
        va_tmp['score'] = oof_pred[va_idx]
        fold_auc = session_auc(va_tmp, 'session_id', target_col, 'score')
        print(f"[{target_col}] Fold {fold} session-AUC: {fold_auc:.4f}")

    oof_df = train_df[['session_id', target_col]].copy()
    oof_df['score'] = oof_pred
    auc_final = session_auc(oof_df, 'session_id', target_col, 'score')
    print(f"[{target_col}] CV session-AUC: {auc_final:.4f}")
    return models, oof_pred, auc_final

click_models, click_oof, click_auc = fit_lgbm(train_df, 'clicked')
order_models, order_oof, order_auc = fit_lgbm(train_df, 'ordered')


[LightGBM] [Info] Number of positive: 79535, number of negative: 2139509
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 7116
[LightGBM] [Info] Number of data points in the train set: 2219044, number of used features: 33
[LightGBM] [Info] Using requested OpenCL platform 0 device 0
[LightGBM] [Info] Using GPU Device: NVIDIA A100-SXM4-40GB, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 31 dense feature groups (67.72 MB) transferred to GPU in 0.078734 secs. 1 sparse feature groups
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.035842 -> initscore=-3.292134
[LightGBM] [Info] Start training from score -3.292134
[clicked] Fold 0 session-AUC: 0.6384
[LightGBM] [Info] Number of positive: 79312, number of negative: 2139732
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 7117
[LightGBM] [Info

In [ ]:
test_df, _, _ = prep(test_raw, is_train=False)

def predict_avg(models, X):
    preds = [m.predict_proba(X)[:,1] for m in models]
    return np.mean(preds, axis=0)

test_df['p_click'] = predict_avg(click_models, test_df[FEATURES])
test_df['p_order'] = predict_avg(order_models, test_df[FEATURES])

W_ORDER, W_CLICK = 0.7, 0.3
test_df['final_score'] = W_ORDER*test_df['p_order'] + W_CLICK*test_df['p_click']


In [ ]:
def make_submission(df, session_col='session_id', item_col='content_id_hashed', score_col='final_score'):
    rows = []
    for sid, g in df.groupby(session_col):
        g = g.sort_values(score_col, ascending=False)
        rows.append((sid, " ".join(g[item_col].astype(str).tolist())))
    return pd.DataFrame(rows, columns=['session_id','prediction'])

submission = make_submission(test_df)
submission.to_csv("submission.csv", index=False)
submission.head()


,session_id,prediction
0,test_0001ff614df60933,12adbd7e465b4adb 8565816477ed3fc5 d13c66f77566...
1,test_00041895a35c4813,8f2de3f4559ae502 39ad83a925c864b2 e95b82f98246...
2,test_00058d4dc9727758,532f78e69e290328 fe92918996a556d8 df2e86f22324...
3,test_00093c8fc0637123,fad649395395c204 c76cba3b657ab7a5 d57d8b42591b...
4,test_000a319a9cf10e66,7189b8cabd825638 e942b8e406c8b298 475615e6fe41...


In [ ]:
# =========================
# LightGBM Feature Importance (avg over folds)
# =========================
def get_lgb_importance(lgb_models, feature_names, importance_type="gain"):
    """
    lgb_models: LGBMClassifier listesi (fold başına bir model)
    feature_names: FEATURES listesi
    importance_type: 'gain' veya 'split' (booster_ için)
    """
    imps = []
    for m in lgb_models:
        if hasattr(m, "feature_importances_"):
            # sklearn API
            imps.append(m.feature_importances_)
        elif hasattr(m, "booster_"):
            # native booster
            imps.append(m.booster_.feature_importance(importance_type=importance_type))
        else:
            # nadiren: native model objesi
            imps.append(m.feature_importance(importance_type=importance_type))

    imps = np.vstack(imps)
    mean_imp = imps.mean(axis=0)
    std_imp  = imps.std(axis=0)

    df = (pd.DataFrame({
            "feature": feature_names,
            "lgb_importance_mean": mean_imp,
            "lgb_importance_std": std_imp
          })
          .sort_values("lgb_importance_mean", ascending=False)
          .reset_index(drop=True))
    df["lgb_importance_rank"] = df["lgb_importance_mean"].rank(ascending=False, method="dense").astype(int)
    return df

# Click modeli için
click_importance = get_lgb_importance(click_models, FEATURES, importance_type="gain")
print("\nTop 10 Önemli Özellikler (Click - LightGBM):")
print(click_importance.head(44)[["feature","lgb_importance_mean"]])

# Order modeli için
order_importance = get_lgb_importance(order_models, FEATURES, importance_type="gain")
print("\nTop 10 Önemli Özellikler (Order - LightGBM):")
print(order_importance.head(44)[["feature","lgb_importance_mean"]])

# (İstersen CSV'ye yaz)
# click_importance.to_csv("fi_click_lgb.csv", index=False)
# order_importance.to_csv("fi_order_lgb.csv", index=False)



Top 10 Önemli Özellikler (Click - LightGBM):
                         feature  lgb_importance_mean
0            user_tenure_in_days               7569.8
1                total_cart_mean               7246.0
2                 total_fav_mean               6839.0
3               total_order_mean               6193.4
4                 original_price               4941.8
5      leaf_category_click_count               4469.2
6                       term_ctr               4243.4
7               product_age_days               4053.8
8                            ctr               3616.8
9                  u_click_ratio               3234.8
10              content_rate_avg               3134.6
11                   content_ctr               2874.0
12       total_search_impression               2710.8
13                   u_conv_rate               2506.8
14                  u_cart2order               2490.2
15              popularity_score               2362.8
16           visual_review_ratio    